Add setCard to STPPaymentCardTextField

benzguo opened the pull request:
Fixes #219 

8957a9b adds some testing UI to `PaymentViewController` – planning on removing the commit when this is good to go.

jflinter commented on Stripe/PublicHeaders/UI/STPPaymentCardTextField.h:
```diff
@@ -220,6 +220,11 @@
  */
 @property(nonatomic, readonly, nullable) STPCardParams *card;
```
This should probably just update the property to be readwrite, rather than defining a separate method.

benzguo replied:
ah, yup.

benzguo replied:
actually, forgot there was a reason I did this – we probably don't want to change the `card` on the legacy `PTKView : STPPaymentCardTextField` interface to `readwrite`, and we can't update this property without also updating that subclass.

jflinter replied:
I think it's fine to do that; that thing is deprecated anyway, and this would be an additive change.

benzguo replied:
:+1:

jflinter commented on Stripe/UI/STPPaymentCardTextField.m:
```diff
@@ -403,6 +403,55 @@ - (STPCardParams *)card {
     return c;
 }
 
+- (void)setCard:(nonnull STPCardParams *)card {
```
This is kind of a lot of hard-to-reason-about code. Could you add some tests?

benzguo replied:
:+1: added tests in https://github.com/stripe/stripe-ios/commit/003d702e001e99b8ccac0b0a7c501c81b8e97a9a

benzguo replied:
Note that these tests don't cover the cases when the field is first responder – we'd need to add an application test target. (Happy to do that if you think we should).

jflinter replied:
What do you think is best?

benzguo replied:
Moar tests! I'll add a new test target.

benzguo commented:
Note: I've removed the testing UI from this branch. It's still available on https://github.com/stripe/stripe-ios/tree/bg-setcardparams-testing if you want to play around.

jflinter commented on Stripe/UI/STPPaymentCardTextField.m:
```diff
@@ -403,6 +403,55 @@ - (STPCardParams *)card {
 }
 
+- (void)setCard:(nonnull STPCardParams *)card {
+    if (card.number.length) {
```
Won't this break if you pass e.g. a card with an empty number to a field where the number is already set? (same with expiration and cvc below.)

jflinter replied:
Like, ideally you could reset some of these values by passing an empty card object.

benzguo replied:
Ah, good point – I'll update.

benzguo commented:
re-r? @jflinter 

updates:
- added an application test target (hosted by the Simple example), and added the new tests to CI.
- `card` is now a `readwrite` property
- setting a `card` with empty fields now behaves correctly

OhDavit commented:
why initially is implemented as a readonly?
